docs: Added a page on pin control for ZMK (#2508)
zmkfirmware/zmk: docs/docs/development/hardware-integration/pinctrl.mdx

:::

A basic keyboard design as introduced in the [new shield guide](./new-shield.mdx) only uses its pins for the keyboard matrix. Many keyboard designs make use of advanced components or functionality, such as displays or shift registers. This results in the keyboard making use of communication protocols such as (but not limited to) SPI, I2C, or UART. Configuring pins for the usage of advanced functionality such as drivers for the previously named protocols is referred to as "Pin Control".

:::warning
The details of pin control can vary from vendor to vendor. An attempt was made to be as general as possible, but it isn't possible to cover all possible cases. The approaches for the nRF52840 and RP2040 MCUs/SoCs are documented in their entirety below. For other MCUs/SoCs, please refer to the [Zephyr documentation](https://docs.zephyrproject.org/3.5.0/index.html) and the examples and other files found in-tree of [ZMK](https://github.com/zmkfirmware/zmk/tree/main/app/boards) and [ZMK's fork of Zephyr](https://github.com/zmkfirmware/zephyr).
:::

## Boards, Shields, and Modules

> :::tip

Reviewer: Do we need this tip box anymore? Or can we move its contents to under the "Predefined Nodes" section?

If this is the case, then you should fork the source repository and add the overlay to the fork. Use said fork to build your firmware, and potentially submit a PR to upstream.
:::

## Predefined Nodes

Many boards will already have some pins configured for particular protocols. We recommend using these pins whenever possible.

<InterconnectTabs items={Metadata} gpio={false} />

Note that some boards may have these pins configured to fit the above pinouts, rather than the capabilities of the MCU. Most notably, boards using the nRF52840 often end up putting low-frequency pins at the locations of the SPI buses above. More on low-frequency pins in the next section.

## Pin Selection

Prior to setting up pin control, you will want to make sure that the pins you use are suitable for the driver that you are using. For this purpose, you will most likely need to consult the datasheet of your MCU/SoC directly. Different MCUs/SoCs have different restrictions in regards to this. The nRF52840 and RP2040 are excellent examples:

- The nRF52840 has its pins marked as either low frequency or high frequency.

>   - Low frequency pins are limited to a frequency of 10KHz, making them unusable for I2C, UART, and (most of) SPI.

Reviewer: ```suggestion
  - Low frequency pins should not be used with signals with frequencies exceeding 10KHz, so they are not ideal for I2C, UART, and (most of) SPI.
```
I think it'd be better to document what happens when you do use low frequency pins (interference with radio) rather than say don't use them, given that many designs seem to be using them anyway. Would you agree?

Note that not all peripherals require all pins to be defined - for example, there are plenty of SPI devices which do not require either the MISO (Main In, Sub Out) or the MOSI (Main Out, Sub In) pin.

## Pin Control File

Pin control consists of two parts:

1. Defining and grouping the pins together for a driver
2. Augmenting a driver/bus node and assigning pin groups to it

> It is standard to do this in two separate files. Create a file called `<my-board>-pinctrl.dtsi`, which will be used for the first part. This file will later be imported into your board's `.dts`. To avoid namespace collisions, if you are configuring pin control for a [shield](#boards-shields-and-modules), then you should write the contents of both parts into the `.overlay` file directly.

Reviewer: Why would there be namespace collisions? (You'd do a relative import, which AFAIK shouldn't have issues)

Author: You're right, this is from when I thought my idea to edit existing shields would work and if it did then there would have been namespace collisions.

The following pin properties can be assigned to groups:

- `bias-disable`: Disable pull-up/down (default, not required).
- `bias-pull-up`: Enable pull-up resistor.
- `bias-pull-down`: Enable pull-down resistor.
- `input-enable`: Enable input from the pin.
- `input-schmitt-enable`: Enable input hysteresis.
- `drive-strength`: Set the drive strength of the pin, in milliamps. Possible values are: 2, 4, 8, 12 (default: 4mA)
- `slew-rate`: If set to 0, slew rate is set to slow. If set to 1, it is set to fast.

</TabItem>
</Tabs>

## Driver/Bus Node

Once pin control for a driver/bus has been defined, you'll need to adjust another node defining the driver/bus. This adjustment can be done in a number of places by convention:

- If defining a unique board, `<board>.dts`
- If defining boards with multiple revisions/versions that share pin control, `<board>-common.dtsi` (which is then included by each `<board>_<revision>.dtsi`)
- If [configuring boards for a shield](#boards-shields-and-modules), directly in the `<board>.overlay` file

> You'll want to identify the correct node for you to be changing. The nRF52840 has nodes defined in `dts/arm/nordic/nrf52840.dtsi`, while the RP2040 has nodes defined in `dts/arm/rpi_pico/rp2040.dtsi`. **Make sure the node that you're using is not already in use elsewhere.**

Reviewer: What is the reason for the "not already in use" remark? I didn't understand it.

Author: Both the rp2040 and the nrf52840 (and I'd imagine most other MCUs) have a limited number of each bus that can be used, even though the pins that can be tied to each bus are flexible. The nrf52 has a maximum of two I2C buses, and xiao ble sense (Zephyr's 3.7 version, not ours, ours isn't set up for this) uses one of these internally for the gyroscope. If you were to use the same node for I2C as the gyroscope uses, then you'd need to override pinctrl and disable the gyroscope. 

While the gyroscope is irrelevant, this might be relevant for a different board or if someone is defining multiple of the same bus on one device.

Reviewer: I see, I interpreted this as more like "don't modify `spi0` if it was already defined" which you will want to do in some/most cases. (There might not be any devices on the bus, or you don't care about them.) It might be better to add a generic warning like "pay attention to what bus you are modifying, which might already have devices defined for it".

```dts
&spi0 {
    status = "okay";
};
```

You would then want to make any adjustments to the node that are necessary, for example adjusting the clock speed. See the Zephyr API documentation for your `compatible` property to see the available properties for customisation. It is recommended to read through the [description of important properties](https://docs.zephyrproject.org/3.5.0/build/dts/intro-syntax-structure.html#dt-important-props), potentially with the addition of [this blog post](https://interrupt.memfault.com/blog/practical_zephyr_dt#zephyrs-dts-skeleton-and-addressing) if `#address-cells` is confusing you.

For SPI specifically, you would create a child node within your SPI bus for each device making use of the SPI bus.

```dts
&spi0 {
    cs-gpios = <&gpio0 15 GPIO_ACTIVE_LOW>, <&gpio0 17 GPIO_ACTIVE_LOW>;
    device1: device@0 {
        compatible = "manufacturer,device";
        reg = <0>;
        spi-max-frequency = <1000000>; /* conservatively set to 1MHz */
    };
    device2: device@1 {
        compatible = "manufacturer,device";
        reg = <1>;
        spi-max-frequency = <1000000>; /* conservatively set to 1MHz */
    };
};
```

> Additional information on configuring specific SPI buses can be found in other pages of the ZMK documentation, or in the Zephyr documentation.

Reviewer: "other pages of the ZMK documentation" is this true? 😄 I can't think of a page off the top of my head where this is documented.

Reviewer: Maybe you have the shift register docs in mind?

Author: So in general, the content on this page is focused on setting up pinctrl and in doing so the bus for devices such as shift registers, RGB, displays, etc. The content on setting up the child node of the bus should be in their respective pages, which is what I'm referring to here. I'll rephrase to make it a bit clearer.

Reviewer: That makes sense, OK.

---
title: Board Pin Control
sidebar_label: Pin Control
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

:::info
This page exists to provide a guide to [Pin Control](https://docs.zephyrproject.org/3.5.0/hardware/pinctrl/index.html#pin-control) for ZMK users and designers. Refer to [Zephyr's page on Pin Control](https://docs.zephyrproject.org/3.5.0/hardware/pinctrl/index.html#pin-control) for elaboration and more details on any of the points raised here.
:::

A basic keyboard design as introduced in the [new shield guide](./new-shield.mdx) only uses its pins for the keyboard matrix. Many keyboard designs make use of advanced components or functionality, such as displays or shift registers. This results in the keyboard making use of communication protocols such as (but not limited to) SPI, I2C, or UART. Configuring pins for the usage of advanced functionality such as drivers for the previously named protocols is referred to as "Pin Control".

:::warning

> The details of pin control can vary from vendor to vendor. An attempt was made to be as general as possible, but it isn't possible to cover all possible cases. The approaches for the nRF52840 and rp2040 MCUs/SoCs are documented in their entirety below. For other MCUs/SoCs, please refer to the [Zephyr documentation](https://docs.zephyrproject.org/3.5.0/index.html) and the examples and other files found in-tree of [ZMK](https://github.com/zmkfirmware/zmk/tree/main/app/boards) and [ZMK's fork of Zephyr](https://github.com/zmkfirmware/zephyr).

Reviewer: I think these links also apply to finding an existing pinctrl node/bus for a given board, assuming the user is writing a shield. It might be good info, given e.g. Xiao buses are defined like you note below.

Author: While true, I think that would be overloading this block a bit. Not sure on best way to handle.

Reviewer: I realized that these are actually documented for the defined interconnects at 
https://deploy-preview-2508--zmk.netlify.app/docs/development/hardware-integration/new-shield#shield-overlays already. This might be a good opportunity to copy those images here, along with the bus node names as documented. Maybe it can be a separate subsection like "Existing pinctrl definitions for interconnects" (draft name) that replaces the tip box right below?

Author: Good idea, have done so. Had to mess with the interconnect tabs bit somewhat, but it works.

:::

## Boards, Shields, and Modules

:::tip
Some boards, such as the Seeed Studio Xiao, already have some of their pins configured for particular communication protocols. If this is the case, then you can use the pre-configured bus rather than configuring one manually (you merely need to identify the correct label for the bus node).
:::

Pin control is always defined for a _board_, never for a shield:

- If you are defining a keyboard that is a _board_, then you will be editing and adding on to the files in your keyboard's folder as usual.
- If your keyboard consists of a _board and a shield_, then you should define a new folder inside of your shield's folder called `boards`. You will need to add a `<board>.overlay` for each board to be used with said shield. Take for example the file structure of the "sofle" shield:
  ```plaintext
  boards/shields/sofle/
  ├── boards/
  │   ├── nice_nano.overlay
  │   ├── nice_nano_v2.overlay
  │   ├── nrfmicro_11.overlay
  │   └── nrfmicro_13.overlay
  └── <sofle shield defining files>
  ```
  Note that you will need to define a separate overlay _for each_ of the boards to be used with the shield.

> ### External Shields

Reviewer: How about we pose this section in a different way, to side step the issue: We could call this section "adding pinctrl to existing keyboards". This way, we can come at this from a user perspective where things don't have to be 100% compliant, where we can then e.g. recommend putting it in `config/shield.overlay` (I think that should work).

In such a case I also don't know if we'd recommend a PR; they could simply be bodging a display or something to their keyboards.

Author: Hmmm. I think we end up wanting both, though from past experience I think we're better off putting it in `config/shield.keymap`. If people want to configure the standard bus then they should PR, otherwise they should `.keymap`.

Reviewer: The keymap is better because it overrides everything, but it isn't per side (and it seems bad to recommend the per-part keymap trick). The overlay is sometimes overridden by upstream but if the node wasn't defined at all it shouldn't be an issue?

Author: It turns out that my idea doesn't work after all, so I removed the section and added the forking note.

Note that not all peripherals require all pins to be defined - for example, there are plenty of SPI devices which do not require either the MISO (Main In, Sub Out) or the MOSI (Main Out, Sub In) pin.

## Pin Control File

Pin control consists of two parts:

1. Defining and grouping the pins together for a driver

> 2. Defining a driver node and assigning pin groups to it

Reviewer: "Bus" sounds more familiar to me than "driver", I see you used both below. Maybe add bus here similarly?

One note, for 2. is "overriding/augmenting" rather than "defining" better, since MCU/SoC defines the bus?

MCUs/SoCs will also have pinctrl header files found under `zephyr/include/zephyr/dt-bindings/pinctrl/`. You will want to make sure that the header file corresponding to your MCU/SoC is imported into your board's `.dts`/`.overlay`. For the nRF52840, your board should already be importing a file such as `nordic/nrf52840_qiaa.dtsi`, which includes the pinctrl header via the following inclusions:

```plaintext
nrf-pinctrl.h -> nrf_common.dtsi -> nrf52840.dtsi -> nrf52840_qiaa.dtsi
```

The corresponding file for the rp2040 (`rpi_pico/rp2040.dtsi`), however, does not include the pinctrl header. Hence you will need to import said header at the top of your `<my-board>-pinctrl.dtsi` file:

```dts title="<my-board>-pinctrl.dtsi"
#include <dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h>
```

If you are configuring pin control for a [shield](#boards-shields-and-modules), then this file may already be imported by the board's definition, in which case you shouldn't re-include it here.

:::info
Some MCUs/SoCs may have their headers located in a Zephyr HAL module. See for example the [NXP HAL module](https://github.com/zephyrproject-rtos/hal_nxp/tree/master/dts/nxp/nxp_imx/rt).

> :::

Reviewer 2: Maybe it's better to use the SAMD21 HAL as an example? I think more people will be developing boards using the samd21 rather than the teensy, although admittedly it has some complications around it's SERCOM blocks rather than discrete spi,i2c etc blocks. The imx pinctrl is also rather complex in its own way

Author: I have no particular opinion either way, that change has been made.

&pinctrl {
    /* your modifications go here */
};
```

Within said node, you will configure one or more child nodes for the buses. You will want to define the child nodes according to the instructions in the `pinctrl.yaml` file.
The child nodes that you define should be named appropriately. The common naming schema is `usageNumber_state`. For example, `uart0_default`.

Child nodes are (generally, there are[exceptions](https://docs.zephyrproject.org/3.5.0/hardware/pinctrl/index.html#pin-configuration)) expected to contain one or more subnodes typically named "groupX". These are for grouping together pins that should be assigned the same state, such as enabling an internal pull-up.
Below are some examples of SPI child nodes for the nRF52840 and the rp2040. Further examples are contained within the comments of the respecting `pinctrl.yaml` files.

<Tabs
defaultValue="nrf52840"
values={[
{label: 'nRF52840', value: 'nrf52840'},
{label: 'rp2040', value: 'rp2040'},
]}>
<TabItem value="nrf52840">
```dts title="<board>-pinctrl.dtsi"
&pinctrl {
    /* configuration for spi0 device, default state */
    spi0_default: spi0_default {
        /* group 1 (name is arbitrary) */
        group1 {

>             /* main role: configure P0.1 as SPI clock, P0.2 as SPI MOSI, P0.3 as SPI MISO */

Reviewer: ```suggestion
            /* main role: configure P0.01 as SPI clock, P0.02 as SPI MOSI, P0.03 as SPI MISO */
```

            psels = <NRF_PSEL(SPIM_SCK, 0, 1)>,
                    <NRF_PSEL(SPIM_MOSI, 0, 2)>,
                    <NRF_PSEL(SPIM_MISO, 0, 3)>;
        };
    };

    /* configuration for spi0 device, sleep state */
    spi0_sleep: spi0_sleep {
        /* group 1 */
        group1 {

>             /* main role: configure P0.1 as SPI clock, P0.2 as SPI MOSI, P0.3 as SPI MISO */

Reviewer: ```suggestion
            /* main role: configure P0.01 as SPI clock, P0.02 as SPI MOSI, P0.03 as SPI MISO */
```

            psels = <NRF_PSEL(SPIM_SCK, 0, 1)>,
                    <NRF_PSEL(SPIM_MOSI, 0, 2)>,
                    <NRF_PSEL(SPIM_MISO, 0, 3)>;
            low-power-enable;
        };
    };

};

````
Pins are always selected via assignments to `psels`. `NRF_PSEL` is a helper macro with the following arguments:
- Pin function configuration. This is the `name` portion of one of the `NRF_FUNC_{name}` macros found in `zephyr/include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h`.
- Port (0 or 1).

> - Pin (0..31).

Reviewer: ```suggestion
- Pin (0 to 31).
```

            pinmux = <SPI0_SCK_P18>, <SPI0_TX_P19>;
        };
        /* group 2 */
        group2 {
            /* configure P16 as SPI0 MISO */
            pinmux = <SPI0_RX_P16>;
            /* enable input on pin 1 */
            input-enable;
        };
    };

    /* configuration for an spi device using PIO0, default state  */
    pio0_spi_default: pio0_spi_default {
        /* group 1 */
        group1 {
            /* Configure P13 and P14 for PIO, to be used for MOSI and SCK */
            pinmux = <PIO0_P13>, <PIO0_P14>;
        };
        /* group 2 */
        group2 {
            /* Configure P16 for PIO to be used for MISO with input */
            pinmux = <PIO0_P12>;
            input-enable;
        };
    };
};
````

The values that can be listed in `pinmux` to assign functionality to pins are listed in `zephyr/include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h`.

The following pin properties can be assigned to groups:

- bias-disable: Disable pull-up/down (default, not required).
- bias-pull-up: Enable pull-up resistor.
- bias-pull-down: Enable pull-down resistor.
- input-enable: Enable input from the pin.

> - input-schmitt-enable: Enable input hysteresys.

Reviewer: ```suggestion
- input-schmitt-enable: Enable input hysteresis.
```
I assume?

- drive-strength: Set the drive strength of the pin, in milliamps. Possible values are: 2, 4, 8, 12 (default: 4mA)

> - slew-rate: If set to 0, slew rate is set to slow. If set to 1, it is set to fast.

Reviewer: Would be good to `code` the property names here.

```dts title="<my-board>-pinctrl.dtsi"
&pinctrl {
    /* your modifications go here */
};
```

Within said node, you will configure one or more child nodes for the buses. You will want to define the child nodes according to the instructions in the `pinctrl.yaml` file.
The child nodes that you define should be named appropriately. The common naming schema is `usageNumber_state`. For example, `uart0_default`.

Child nodes are (generally, there are[exceptions](https://docs.zephyrproject.org/3.5.0/hardware/pinctrl/index.html#pin-configuration)) expected to contain one or more subnodes typically named "groupX". These are for grouping together pins that should be assigned the same state, such as enabling an internal pull-up.
Below are some examples of SPI child nodes for the nRF52840 and the RP2040. Further examples are contained within the comments of the respecting `pinctrl.yaml` files.

<Tabs
defaultValue="nrf52840"
values={[
{label: 'nRF52840', value: 'nrf52840'},
{label: 'RP2040', value: 'rp2040'},
]}>
<TabItem value="nrf52840">
```dts title="<board>-pinctrl.dtsi"
&pinctrl {
    /* configuration for spi0 device, default state */
    spi0_default: spi0_default {

>         /* group 1 (name is arbitrary) */

Reviewer: ```suggestion
        /* node name is arbitrary */
```

        group1 {
            /* main role: configure P0.01 as SPI clock, P0.02 as SPI MOSI, P0.03 as SPI MISO */
            psels = <NRF_PSEL(SPIM_SCK, 0, 1)>,
                    <NRF_PSEL(SPIM_MOSI, 0, 2)>,
                    <NRF_PSEL(SPIM_MISO, 0, 3)>;
        };
    };

    /* configuration for spi0 device, sleep state */
    spi0_sleep: spi0_sleep {

>         /* group 1 */

Reviewer: ```suggestion
```

            psels = <NRF_PSEL(SPIM_SCK, 0, 1)>,
                    <NRF_PSEL(SPIM_MOSI, 0, 2)>,
                    <NRF_PSEL(SPIM_MISO, 0, 3)>;
            low-power-enable;
        };
    };

};

````
Pins are always selected via assignments to `psels`. `NRF_PSEL` is a helper macro with the following arguments:
- Pin function configuration. This is the `name` portion of one of the `NRF_FUNC_{name}` macros found in `zephyr/include/zephyr/dt-bindings/pinctrl/nrf-pinctrl.h`.
- Port (0 or 1).
- Pin (0 to 31).

The following pin properties can be assigned to groups:

- `bias-disable`: Disable pull-up/down (default behavior, not required).
- `bias-pull-up`: Enable pull-up resistor.
- `bias-pull-down`: Enable pull-down resistor.
- `low-power-enable`: Configure pin as an input with input buffer disconnected.

Note that bias options are mutually exclusive.

There is an additional child node for the "sleep" state, configuring the pins for low power. More on states will be explained later.
</TabItem>
<TabItem value="rp2040">
```dts title="<board>-pinctrl.dtsi"
&pinctrl {
    /* configuration for spi0 device, default state */
    spi0_default: spi0_default {

>         /* group 1 */

Reviewer: ```suggestion
```

        group1 {
            /* configure P18 as SPI0 clock, P19 as SPI0 MOSI */
            pinmux = <SPI0_SCK_P18>, <SPI0_TX_P19>;
        };

>         /* group 2 */

Reviewer: ```suggestion
```

        group2 {
            /* configure P16 as SPI0 MISO */
            pinmux = <SPI0_RX_P16>;
            /* enable input on pin 1 */
            input-enable;
        };
    };

    /* configuration for an spi device using PIO0, default state  */
    pio0_spi_default: pio0_spi_default {

>         /* group 1 */

Reviewer: ```suggestion
```

        group1 {
            /* Configure P13 and P14 for PIO, to be used for MOSI and SCK */
            pinmux = <PIO0_P13>, <PIO0_P14>;
        };

>         /* group 2 */

Reviewer: ```suggestion
```

        pinctrl-names = "default";
        compatible = "raspberrypi,pico-spi-pio";
        #address-cells = <1>;
        #size-cells = <0>;
        clocks = <&system_clk>;
        clock-frequency = <4000000>;
        /* These pins should be the same as in pinctrl */
        miso-gpios = <&gpio0 12 0>;
        clk-gpios = <&gpio0 14 GPIO_ACTIVE_HIGH>;
        mosi-gpios = <&gpio0 15 GPIO_ACTIVE_HIGH>;

        cs-gpios = <...>; // List of chip select gpios, one for each device
        /* Nodes using the bus go here */
    };
};
```

Depending on the desired usage for PIO, you will want to adjust the `compatible` property and the SPI-specific properties (`miso-gpios`, `clk-gpios`, `mosi-gpios`). See the Zephyr API documentation for information about alternative PIO drivers. Once defined, SPI can be used via PIO as presented in the previous subsection, referring to `pio0_spi` (or similar) instead of `spi0`.

### Additional examples

Below are examples for UART and I2C, as the other two most common usages for pin control.

#### UART nRF52840

In the pin control file:

```dts
&pinctrl {
    /* configuration for uart0 device, default state */
    uart0_default: uart0_default {

>         /* group 1 */

Reviewer: ```suggestion
```

        group1 {
            /* configure P0.1 as UART_TX and P0.2 as UART_RTS */
            psels = <NRF_PSEL(UART_TX, 0, 1)>, <NRF_PSEL(UART_RTS, 0, 2)>;
        };

>         /* group 2 */

Reviewer: ```suggestion
```

        group2 {
            /* configure P0.3 as UART_RX and P0.4 as UART_CTS */
            psels = <NRF_PSEL(UART_RX, 0, 3)>, <NRF_PSEL(UART_CTS, 0, 4)>;
            /* both P0.3 and P0.4 are configured with pull-up */
            bias-pull-up;
        };
    };
};
```

In the main file:

```dts
#include "<board>-pinctrl.dtsi"

&uart0 {
    pinctrl-0 = <&uart0_default>;
    pinctrl-names = "default";
};
```

#### UART rp2040

In the pin control file:

```dts
#include <dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h>

&pinctrl {
    /* configuration for the usart0 "default" state */
    uart0_default: uart0_default {

>         /* group 1 */

Reviewer: ```suggestion
```
I feel like these are unnecessary if they don't give additional info.

        group1 {
            /* configure P0 as UART0 TX */
            pinmux = <UART0_TX_P0>;
            };

>         /* group 2 */

Reviewer: ```suggestion
```

        group2 {
            /* configure P1 as UART0 RX */
            pinmux = <UART0_RX_P1>;
            /* enable input on pin 1 */
            input-enable;
        };
    };
};
```

In the main file:

```dts
#include "<board>-pinctrl.dtsi"

&uart0 {
    pinctrl-0 = <&uart0_default>;
    pinctrl-names = "default";
};
```

> #### UART rp2040 PIO

Reviewer: ```suggestion
#### UART RP2040 PIO
```

In the main file:

```dts
#include "<board>-pinctrl.dtsi"

&pio0 {
    status = "okay";

    pio0_uart: serial {
        status = "okay";
        compatible = "raspberrypi,pico-uart-pio";
        pinctrl-0 = <&pio0_uart_default>;
        pinctrl-names = "default";
        current-speed = <115200>;
    };
};
```

#### I2C nRF52840

Specifically for an I2C controller, aka Nordic TWIM.

In the pin control file:

```dts
&pinctrl {
    /* configuration for i2c0 device, default state */
    i2c0_default: i2c0_default {
        group1 {
            psels = <NRF_PSEL(TWIM_SDA, 0, 7)>,
                <NRF_PSEL(TWIM_SCL, 0, 27)>;
        };
    };

    i2c0_sleep: i2c0_sleep {
        group1 {
            psels = <NRF_PSEL(TWIM_SDA, 0, 7)>,
                <NRF_PSEL(TWIM_SCL, 0, 27)>;
            low-power-enable;
        };
    };
};
```

In the main file:

```dts
#include "<board>-pinctrl.dtsi"

&i2c0 {
    compatible = "nordic,nrf-twim"; // I2C controller instead of generic
    status = "okay";
    pinctrl-0 = <&i2c0_default>;
    pinctrl-1 = <&i2c0_sleep>;
    pinctrl-names = "default", "sleep";
    clock-frequency = <I2C_BITRATE_FAST>;

    /* Nodes using the bus go here */
};
```

> #### I2C rp2040

Reviewer: ```suggestion
#### I2C RP2040
```

In the pin control file:

```dts
#include <dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h>

&pinctrl {
    /* configuration for the i2c0 "default" state */
    i2c0_default: i2c0_default {
        group1 {
            pinmux = <I2C0_SDA_P4>, <I2C0_SCL_P5>;
            input-enable;
            input-schmitt-enable;
        };
    };
};
```

In the main file:

```dts
#include "<board>-pinctrl.dtsi"

&i2c0 {
    status = "okay";
    pinctrl-0 = <&i2c0_default>;
    pinctrl-names = "default";
    clock-frequency = <I2C_BITRATE_STANDARD>;

    /* Nodes using the bus go here */
};
```

> #### I2C rp2040 PIO

Reviewer: ```suggestion
#### I2C RP2040 PIO
```

> Zephyr currently does not have support for I2C using rp2040 PIO.

Reviewer: ```suggestion
Zephyr currently does not have support for I2C using RP2040 PIO.
```
